torch/csrc/distributed/c10d/init.cpp

```diff
@@ -106,6 +106,28 @@ class PythonStore : public ::c10d::Store {
   }
 };
 
+// PythonhookBinder has a static function called register_comm_hook
+// that enables registering a c10d PythonCommHook object to c10d
+// reducer from Python.
+class PythonHookBinder {
+ public:
+  // This static method is called from DDP's Python API. Its inputs are
+  // a c10d reducer object, state, and callable comm_hook. State and
+  // comm_hook inputs are Python objects and this function creates a
+  // c10d PythonCommHook object using these inputs. It later calls
+  // register_comm_hook function of the reducer input to register that
+  // PythonCommHook object.
+  static void register_comm_hook(
+      py::object& ddp_model,
```

Comment:
Pass `Reducer` as a parameter here and call `register_comm_hook` on the Reducer.

Reply:
@pritamdamania I had comment on passing ddp_model instead of reducer here, because reducer is quite internal thing and has dependencies on ddp model.  another benefit to pass model is: users only need to wrap ddp model and pass it here, do not need to worry about create reducer and pass it in

Reply:
@zhaojuanmao I didn't mean we should expose the Reducer to users. We should have only one public API for registering the hook and it should be something like `ddp._register_comm_hook(state, comm_hook)` (be a method of the DistributedDataParallel class). Then the python implementation of `_register_comm_hook`, would call the C++ method as follows: `dist._register_comm_hook(self.reducer, state, comm_hook)`.

Reply:
Makes sense to me. If all agree, I'll revert it back to smth like this:
https://github.com/pytorch/pytorch/blob/ff7587df9521a0cd4c85ac34c6e59912ce2c8c3e/torch/csrc/distributed/c10d/init.cpp#L120-L126